discovery/digitalocean/digitalocean.go

```diff
@@ -0,0 +1,243 @@
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client    *godo.Client
+	tag       string
+	port      int
+	separator string
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		tag:       conf.Tag,
+		port:      conf.Port,
+		separator: conf.Separator,
+	}
+
+	var rt http.RoundTripper = &http.Transport{
+		IdleConnTimeout: 5 * time.Duration(conf.RefreshInterval),
+		DialContext: conntrack.NewDialContextFunc(
+			conntrack.DialWithTracing(),
+			conntrack.DialWithName("digitalocean_sd"),
+		),
+	}
+	if conf.AccessToken != "" {
+		rt = config_util.NewBearerAuthRoundTripper(conf.AccessToken, rt)
+	}
+	if conf.AccessTokenFile != "" {
+		rt = config_util.NewBearerAuthFileRoundTripper(conf.AccessTokenFile, rt)
+	}
+
+	var err error
+	d.client, err = godo.New(
+		&http.Client{
+			Transport: rt,
+			Timeout:   5 * time.Duration(conf.RefreshInterval),
+		},
+		godo.SetUserAgent(fmt.Sprintf("Prometheus/%s", version.Version)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up digital ocean agent: %w", err)
+	}
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"digitalocean",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	source := "DO"
+	if d.tag != "" {
+		source += "_" + d.tag
+	}
+	tg := &targetgroup.Group{
+		Source: source,
+	}
+
+	droplets, err := d.listDroplets()
+	if err != nil {
+		return nil, err
+	}
+	for _, droplet := range droplets {
+		if droplet.Networks == nil || len(droplet.Networks.V4) == 0 {
+			continue
+		}
+
+		privateIPv4, err := droplet.PrivateIPv4()
+		if err != nil {
+			return nil, fmt.Errorf("error while reading private IPv4 of droplet %d: %w", droplet.ID, err)
+		}
+		publicIPv4, err := droplet.PublicIPv4()
+		if err != nil {
+			return nil, fmt.Errorf("error while reading public IPv4 of droplet %d: %w", droplet.ID, err)
+		}
+		publicIPv6, err := droplet.PublicIPv6()
+		if err != nil {
+			return nil, fmt.Errorf("error while reading public IPv6 of droplet %d: %w", droplet.ID, err)
+		}
+
+		labels := model.LabelSet{
+			doLabelID:          model.LabelValue(fmt.Sprintf("%d", droplet.ID)),
+			doLabelName:        model.LabelValue(droplet.Name),
+			doLabelImage:       model.LabelValue(droplet.Image.Name),
+			doLabelPrivateIPv4: model.LabelValue(privateIPv4),
+			doLabelPublicIPv4:  model.LabelValue(publicIPv4),
+			doLabelPublicIPv6:  model.LabelValue(publicIPv6),
+			doLabelRegion:      model.LabelValue(droplet.Region.Name),
+			doLabelSize:        model.LabelValue(droplet.SizeSlug),
+			doLabelStatus:      model.LabelValue(droplet.Status),
+		}
+
+		var addr string
+		if privateIPv4 != "" {
+			addr = privateIPv4
+		} else if publicIPv4 != "" {
+			addr = publicIPv4
+		} else {
```

Comment:
Why not v6?

Reply:
By default DigitalOcean provides IPv4 and it is not something you can disable. I prefer consistency.

Reply:
Yes, this should probably check for v6 first, and use v4 as a fallback.

Reply:
done

Reply:
Hmm, is it safe to presume that if v6 is on a droplet that a Prometheus that could get to it via v4 could also get to it via v6?